File: 05-core-bank/REGENERA-05-CORE-BANKING-10-10/src/main/kotlin/com/regenera/core/Idempotency.kt
package com.regenera.core

import java.time.Instant

enum class IdempotencyState {
    RECEIVED,
    PROCESSING,
    COMPLETED,
    FAILED_RETRYABLE,
    FAILED_FINAL,
    UNKNOWN,
    EXPIRED,
}

data class IdempotencyRecord(
    val scope: String,
    val key: String,
    val payloadHash: String,
    val state: IdempotencyState,
    val responseReference: String?,
    val updatedAt: Instant,
)

sealed interface IdempotencyBegin {
    data class Acquired(val record: IdempotencyRecord) : IdempotencyBegin
    data class Replay(val responseReference: String) : IdempotencyBegin
    data class Blocked(val state: IdempotencyState) : IdempotencyBegin
}

class IdempotencyRegistry {
    private val records = linkedMapOf<Pair<String, String>, IdempotencyRecord>()

    @Synchronized
    fun begin(
        scope: String,
        key: String,
        payloadHash: String,
        now: Instant,
    ): IdempotencyBegin {
        validate(scope, key, payloadHash)
        val index = scope to key
        val existing = records[index]

        if (existing == null) {
            val record = IdempotencyRecord(
                scope = scope,
                key = key,
                payloadHash = payloadHash,
                state = IdempotencyState.PROCESSING,
                responseReference = null,
                updatedAt = now,
            )
            records[index] = record
            return IdempotencyBegin.Acquired(record)
        }

        if (existing.payloadHash != payloadHash) {
            throw ConflictException(
                "IDEMPOTENCY_PAYLOAD_CONFLICT",
                "A chave já foi usada com outro payload",
            )
        }

        return when (existing.state) {
            IdempotencyState.COMPLETED -> IdempotencyBegin.Replay(
                existing.responseReference
                    ?: throw IllegalStateException("Resposta concluída sem referência"),
            )
            IdempotencyState.FAILED_RETRYABLE,
            IdempotencyState.EXPIRED,
            -> {
                val reacquired = existing.copy(
                    state = IdempotencyState.PROCESSING,
                    responseReference = null,
                    updatedAt = now,
                )
                records[index] = reacquired
                IdempotencyBegin.Acquired(reacquired)
            }
            else -> IdempotencyBegin.Blocked(existing.state)
        }
    }

    @Synchronized
    fun complete(
        scope: String,
        key: String,
        responseReference: String,
        now: Instant,
    ): IdempotencyRecord = transition(
        scope,
        key,
        allowed = setOf(IdempotencyState.PROCESSING),
        target = IdempotencyState.COMPLETED,
        responseReference = responseReference,
        now = now,
    )

    @Synchronized
    fun markUnknown(scope: String, key: String, now: Instant): IdempotencyRecord =
        transition(
            scope,
            key,
            allowed = setOf(IdempotencyState.PROCESSING),
            target = IdempotencyState.UNKNOWN,
            responseReference = null,
            now = now,
        )

    @Synchronized
    fun failRetryable(scope: String, key: String, now: Instant): IdempotencyRecord =
        transition(
            scope,
            key,
            allowed = setOf(IdempotencyState.PROCESSING),
            target = IdempotencyState.FAILED_RETRYABLE,
            responseReference = null,
            now = now,
        )

    @Synchronized
    fun failFinal(scope: String, key: String, now: Instant): IdempotencyRecord =
        transition(
            scope,
            key,
            allowed = setOf(IdempotencyState.PROCESSING),
            target = IdempotencyState.FAILED_FINAL,
            responseReference = null,
            now = now,
        )

    @Synchronized
    fun expire(scope: String, key: String, now: Instant): IdempotencyRecord =
        transition(
            scope,
            key,
            allowed = setOf(
                IdempotencyState.RECEIVED,
                IdempotencyState.FAILED_RETRYABLE,
                IdempotencyState.FAILED_FINAL,
            ),
            target = IdempotencyState.EXPIRED,
            responseReference = null,
            now = now,
        )

    @Synchronized
    fun get(scope: String, key: String): IdempotencyRecord? = records[scope to key]

    private fun transition(
        scope: String,
        key: String,
        allowed: Set<IdempotencyState>,
        target: IdempotencyState,
        responseReference: String?,
        now: Instant,
    ): IdempotencyRecord {
        val index = scope to key
        val current = records[index] ?: throw NotFoundException(
            "IDEMPOTENCY_RECORD_NOT_FOUND",
            "Registro de idempotência não encontrado",
        )
        if (current.state !in allowed) {
            throw StateTransitionException(
                "IDEMPOTENCY_TRANSITION_FORBIDDEN",
                "Transição ${current.state} -> $target não permitida",
            )
        }
        val next = current.copy(
            state = target,
            responseReference = responseReference,
            updatedAt = now,
        )
        records[index] = next
        return next
    }

    private fun validate(scope: String, key: String, payloadHash: String) {
        if (!scope.matches(Regex("^[a-z][a-z0-9:_-]{2,79}$"))) {
            throw ValidationException(
                "IDEMPOTENCY_SCOPE_INVALID",
                "Escopo de idempotência inválido",
            )
        }
        if (key.length !in 16..128) {
            throw ValidationException(
                "IDEMPOTENCY_KEY_INVALID",
                "Chave de idempotência deve ter entre 16 e 128 caracteres",
            )
        }
        if (!payloadHash.matches(Regex("^[a-f0-9]{64}$"))) {
            throw ValidationException(
                "IDEMPOTENCY_HASH_INVALID",
                "Hash de payload inválido",
            )
        }
    }
}

File: 14-operations/REGENERA-14-OPERATIONS-10-10/src/regenera_operations/idempotency.py
from __future__ import annotations
from dataclasses import dataclass
from enum import Enum
from .errors import ConflictError, StateTransitionError, ValidationError
from .utils import digest

class IdempotencyState(str, Enum):
    PROCESSING = "PROCESSING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    UNKNOWN = "UNKNOWN"

@dataclass(slots=True)
class IdempotencyRecord:
    key: str
    fingerprint: str
    state: IdempotencyState
    result: object | None = None

class IdempotencyRegistry:
    def __init__(self) -> None:
        self._records: dict[str, IdempotencyRecord] = {}

    def begin(self, key: str, payload: object) -> IdempotencyRecord:
        if not key.strip() or len(key) > 128:
            raise ValidationError("chave de idempotência inválida")
        fingerprint = digest(payload)
        current = self._records.get(key)
        if current:
            if current.fingerprint != fingerprint:
                raise ConflictError("chave reutilizada com payload divergente")
            return current
        record = IdempotencyRecord(key, fingerprint, IdempotencyState.PROCESSING)
        self._records[key] = record
        return record

    def complete(self, key: str, result: object) -> IdempotencyRecord:
        record = self._require(key)
        if record.state not in {IdempotencyState.PROCESSING, IdempotencyState.UNKNOWN}:
            raise StateTransitionError("estado não permite conclusão")
        record.state = IdempotencyState.COMPLETED
        record.result = result
        return record

    def fail(self, key: str, result: object | None = None) -> IdempotencyRecord:
        record = self._require(key)
        if record.state != IdempotencyState.PROCESSING:
            raise StateTransitionError("estado não permite falha")
        record.state = IdempotencyState.FAILED
        record.result = result
        return record

    def mark_unknown(self, key: str) -> IdempotencyRecord:
        record = self._require(key)
        if record.state != IdempotencyState.PROCESSING:
            raise StateTransitionError("estado não permite UNKNOWN")
        record.state = IdempotencyState.UNKNOWN
        return record

    def assert_retry_allowed(self, key: str) -> None:
        record = self._require(key)
        if record.state in {IdempotencyState.PROCESSING, IdempotencyState.UNKNOWN, IdempotencyState.COMPLETED}:
            raise StateTransitionError("repetição bloqueada")

    def _require(self, key: str) -> IdempotencyRecord:
        if key not in self._records:
            raise ValidationError("registro de idempotência ausente")
        return self._records[key]
